### DISCUSSION TITLE
Green on Green - Step By Step

### DISCUSSION THREAD

**Discussion by alistairphillipsphx** [geezacoleman/OpenWeedLocator#71] [Q&A]
Has anyone got a tutorial or list of instruction on how to train / implement the new Green on Green function ?

And suggestion greatly received

**Comment by geezacoleman** [comment]
I have put together some steps in the `models` [directory](https://github.com/geezacoleman/OpenWeedLocator/tree/main/models) if that's helpful, though it's not super thorough currently. Which parts would you like to see more detail on?

### Training
For training, you can follow the instructions on the official [Coral colab tutorial](https://colab.research.google.com/github/google-coral/tutorials/blob/master/retrain_ssdlite_mobiledet_qat_tf1.ipynb), alternatively [EdjeElectronics](https://www.youtube.com/watch?v=XZ7FYAMCc4M&ab_channel=EdjeElectronics) also has a great YouTube video and Google Colab tutorial.

Finding datasets of weeds is one of the harder parts. Another project I'm involved with is [Weed-AI](https://weed-ai.sydney.edu.au/about), which has some weeds datasets that may be useful.

Longer term, it would ideal to have the ability to export directly from the YOLOv5/v8 repositories. These models are fairly straightforward to train and are state-of-the-art in performance too. I'm hoping the issues mentioned in this [Issue](https://github.com/ultralytics/ultralytics/issues/1185) will be resolved.

Let me know how it can be improved though or any specifics you'd like detail on and I can try filling in those gaps.

**Comment by geezacoleman** [comment]
I have added a lot more detail now to the [instructions for adding green-on-green](https://github.com/geezacoleman/OpenWeedLocator/tree/main/models). Hopefully this makes the process a little easier!

**Reply by gnaegit** [reply]
**Issue with Pycoral Installation Instructions**

I believe the instruction on [this line](https://github.com/geezacoleman/OpenWeedLocator/blob/main/models/README.md?plain=1#L23) to install Pycoral via `pip` is incorrect. Following that command will install [this package](https://pypi.org/project/pycoral/) from PyPI.

To correctly install Pycoral via `pip`, the following command should be used:

```bash
python3 -m pip install --extra-index-url https://google-coral.github.io/py-repo/ pycoral~=2.0
```

This is the recommended method according to the [Coral documentation](https://coral.ai/docs/accelerator/get-started/#pycoral-on-mac-win). However, please note that only Python versions 3.6-3.9 are supported (this limitation also applies to installation via `apt-get`).

**Comment by david-j-webster** [comment]
having an issue with the coral
please see the screen grab
(owl) pi@raspberrypi:~/owl $ python owl.py --show-display --algorithm gog
[INFO] Setting up nozzles...
[INFO] Nozzle setup complete. Initiating camera...
[INFO] Camera setup complete. Settings: 
Resolution: (416, 320)
Framerate: 40
Exposure Mode: beach
AutoWhiteBalance: auto
Exposure Compensation: -4
Sensor Mode: 0
[WARNING] No model directory or path provided with --model-path flag. Attempting to load from default...
[INFO] Using tf2_ssd_mobilenet_v2_coco17_ptq_edgetpu model...

[ValueError] while starting algorithm: gog.
Error message: Failed to load delegate from libedgetpu.so.1

This is due to an unrecognised Google Coral device. Please make sure it is connected correctly.
If the error persists, try unplugging it and plugging it again or restarting the
Raspberry Pi. For more information visit:
https://github.com/tensorflow/tensorflow/issues/32743

**Reply by geezacoleman** [reply]
Thanks for trying this out. What steps have you taken up to this point? Were you able to install the required drivers ok? If you can provide a few more details I can hopefully help a bit more.

More generally - this error message often comes from not recognising the Google Coral device. Try unplugging it/plugging it in again or restarting the device. Happy to help further though with more details though.

**Reply by david-j-webster** [reply]
i have gone through the step by step into the google page and then completed the other steps... i have ran from scratch to a working focus then the G on G on top    each time i get the same error  unplugged and rebooted nothing .    Thanks for getting back to me so fast.
david

**Reply by geezacoleman** [reply]
Just a few questions to figure this out:
1. Which version Pi are you using?
2. Were there any errors during the GoG installation process and did you make sure the USB was plugged in/out when needed?
3. Which cable are you using for the USB accelerator? Some reports of the official google one being dodgy. Could be worth trying another USB C cable.
4. Have you tried a different USB port on the Pi? See if the USB2.0 or USB3.0 work differently.
5. Do you reboot with it plugged in?

I would recommend checking through some of the responses provided in the tensorflow issue thread [here](https://github.com/tensorflow/tensorflow/issues/32743) to see if any are relevant. Broad themes are basically the error comes from permissions/rebooting/hardware. Do you have a picture of the set up?

**Reply by david-j-webster** [reply]
thanks for your quick reply
i am using a pi4 (would love to use a pi 5, i am also playing with a jetson nano but have not got this far)
i have not had any errors but the one in my previous post (the step by step works well)
i have changed the port from usb3 to usb3 i was not aware that the coral would work on the 2
i have changed the cord and rebooted

i will look through the thread  thanks for the help

**Comment by ElectroS3B** [comment]
Hello, I am a French farmer and I am so impressed by your work. I am going to try to do an OWL machin to detect weed in line crop fields like onion and corn. I have no competence in AI and developpement tools. Would it be possible to detect a crop line (for example a corn line) and only spray the green which is not on this line ? Thanks a lot for your work !
![raies maïs](https://github.com/user-attachments/assets/8bc7b860-ba91-45df-8c54-316431612186)

**Reply by ElectroS3B** [reply]
Another idea would be to make it only recognize the crop, for example corn, and tell it to spray everything green except the crop (we only make it learn the crop at different stages, which limits the size of the database and learning time versus learning all the weeds).

**Reply by geezacoleman** [reply]
Hey @ElectroS3B ! Appreciate your feedback about the project and very exciting to hear you're building a system yourself. Make sure you share progress over on the Community Builds #75.

The short answer to your question is yes (but with a fair bit of work required!) - crop row detection for weed detection is very doable and was one of the first ways people used computer vision for automated weed detection, e.g. [this paper](https://www.sciencedirect.com/science/article/pii/S0168169917303940) but there are many more examples out there. The challenge is running it on the Raspberry Pi in real time and developing something that can reliably find the rows in all conditions. Just as colour-based weed detection is challenging, colour-based crop row detection can be unreliable.

Unless you can develop some sort of shroud and consistent lighting to reliably detect the crop, it's potentially a better idea to train a model to differentiate the crop and weeds as you suggest. There have been [attempts recently](https://www.mdpi.com/2072-4292/16/11/2041) at training on the crop where everything else is a weed with varying levels of success. You'll still need to collect adequate quantities of image data however to get this to work. If you'll be using the OWL, then collecting with the OWL is probably the best way to do it, otherwise a phone camera can serve as a good and easier starting point.

Crop row detection is something I'd like to look at though! Hopefully this answered your questions.
